virtcontainers/kata_agent.go

```diff
@@ -451,6 +451,21 @@ func (k *kataAgent) startSandbox(sandbox *Sandbox) error {
 		hostname = hostname[:maxHostnameLen]
 	}
 
+	//check grpc server is serving
+	if err := k.connect(); err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+	if _, err := k.client.Check(ctx, &grpc.CheckRequest{}); err != nil {
+		k.Logger().Error("grpc server check failed")
+
+		//kill kata-proxy to release resources
+		k.proxy.stop(sandbox, k.state.ProxyPid)
```

Comment:
The rollback of the proxy being started sounds like a good idea. But if we want this to be properly implemented, we should create a `defer` right after the proxy has been correctly started. You need to declare a global `err error` for the function, and if when the defer is executed, the `err != nil`, then we should stop the proxy. This will apply for any error case that could happen in this function after the proxy has been started.
You also want to handle the case where the agent fails to start, which is great, but a separate thing IMO, that's why it should be split into 2 different commits. First you introduce proper rollback for the proxy, and then you introduce the check for the agent availability.